### PR TITLE
Add shared registrations balanced mode

### PR DESCRIPTION
**Pull request by markope**
This is a suggested extra mode for shared registrations.
Shared registrations can be used with mode "balance" in addition to the existing "roundrobin, random, ..." in this pull request.

If a call is requested on the shared URI then crossbar randomly picks one of the registered callees that is not currently busy with an ongoing request. If all registrations are busy then crossbar picks ~~a busy callee randomly~~ the least busy callee by number of pending invocations.

Hope the feature is welcome. I'm happy to discuss...

### CODE REVIEW

**Comment by markope** [comment]
Also there is an AutobahnPython and a crossbar example pull request coming up with related changes. The example also served as the test.

**Comment by dustinlacewell** [comment]
Why not count how many invocations are pending against each observer and sort them?

**Comment by markope** [comment]
ok, thanks. I improved that now.

File: crossbar/router/dealer.py [hunk 1]
```diff
@@ -386,6 +388,18 @@ def on_authorize_success(authorized):
                     elif registration.extra.invoke == message.Register.INVOKE_RANDOM:
                         callee = registration.observers[random.randint(0, len(registration.observers) - 1)]
 
+                    elif registration.extra.invoke == message.Register.INVOKE_BALANCE:
```
Comment (oberstet) [inline crossbar/router/dealer.py]: This code would be a bottleneck ..

**Comment by oberstet** [comment]
see: https://github.com/crossbario/crossbar/issues/606

**Comment by markope** [comment]
what exactly do you mean by 

> ...would be a bottleneck...

Needs improvement or good enough?
Should I adjust anything else apart from the mode name?

**Pull request closed without merging**